aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1315,81 +1315,255 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
           new_indices = indices
-          new_values[k] = values[k].index_select(dim - len(sparse_shape), index) for k in range(nnz)
+          new_values = values.index_select(dim - sparse_dim + 1, index);
     */
-  auto ndim = self.dim();
+  const auto ndim = self.dim();
   if (ndim == 0) {
     TORCH_CHECK_INDEX(false, "index_select() cannot be applied to a 0-dim tensor.");
   }
   if (!(index.dim() == 1 && index.dtype() == at::kLong)) {
     TORCH_CHECK_INDEX(false, "index_select() argument index must be 1-D long-tensor.");
   }
   dim = maybe_wrap_dim(dim, ndim);
-  auto size = self.size(dim);
-  auto sparse_dim = self.sparse_dim();
-  auto dense_dim = self.dense_dim();
-  auto indices = self._indices();
-  auto values = self._values();
-  auto nnz = values.size(0);
-  auto new_sizes = self.sizes().vec();
-  new_sizes[dim] = index.size(0);
-
+  const auto size = self.size(dim);
+  const auto sparse_dim = self.sparse_dim();
+  const auto dense_dim = self.dense_dim();
+  const auto indices = self._indices();
+  const auto values = self._values();
+  const auto nnz = values.size(0);
+  const auto index_len = index.size(0);
+  auto res_sizes = self.sizes().vec();
+  res_sizes[dim] = index_len;
+
+  // If indexing into sparse dimensions
   if (dim < sparse_dim) {
+    auto nneg_index = at::empty_like(index);
+    {
+      auto* ptr_index = index.data_ptr<int64_t>();
+      auto* ptr_index_end = ptr_index + index_len;
+      auto* ptr_nneg_index = nneg_index.data_ptr<int64_t>();
+      while (ptr_index != ptr_index_end) {
+        auto idx = *ptr_index++;
+        if (idx < -size || idx >= size) {
+          TORCH_CHECK_INDEX(false,
+              "index_select(): index contains ", idx, " that is out of range for tensor of size ",
+              self.sizes(), " at dimension ", dim
+          );
+        }
+        if (idx < 0) {
+          idx += size;
+        }
+        *ptr_nneg_index++ = idx;
+      }
+    }
 
-    auto cpu_dim_indices = indices[dim].to(c10::kCPU).contiguous();
-    int64_t* cpu_dim_indices_ptr = cpu_dim_indices.data_ptr<int64_t>();
-    auto cpu_index = index.to(c10::kCPU).contiguous();
-    int64_t* cpu_index_ptr = cpu_index.data_ptr<int64_t>();
-    std::vector<int64_t> zindices;
-    std::vector<int64_t> iindices;
-    int64_t new_nnz = 0;
-    for (const auto i : c10::irange(new_sizes[dim])) {
-      int64_t idx = cpu_index_ptr[i];
-      if (idx < -size || idx >= size) {
-        TORCH_CHECK_INDEX(false, "index_select(): index contains ", idx, " that is out of range for tensor of size ",
-                   self.sizes(), " at dimension ", dim);
+    // Much faster than at::unique_dim
+    const auto unique_with_counts = [](
+        const Tensor& t, int64_t len, bool is_sorted = false
+    ) -> std::tuple<Tensor, Tensor, Tensor, int64_t> {
+      // t_unique will be modified to hold unique values in-place
+      Tensor t_unique, t_idx;
+      if (!is_sorted) {
+        std::tie(t_unique, t_idx) = at::sort(t);
+      }
```

Comment:
We can optimize this part heavily when `nnz > self.sizes(dim)`. We can do a radix sort in such a case.

Reply:
In fact, we can come up with a better algorithm overall when `nnz > self.sizes(dim)`.